### PR TITLE
Goto x

### PR DESCRIPTION
**Pull request by kosivantsov**
## Which ticket is resolved?

- #1641 Navigate between segments inserted from /tm/auto and /tm/enforce
  * https://sourceforge.net/p/omegat/feature-requests/1641/
 

## What does this PR change?

- It introduses 4 new Goto menu items (grouped in a submenu): goto next/prev xauto/xenforced.

### CODE REVIEW

**Comment by brandelune** [comment]
Would you happen to have a test project ?

**Comment by kosivantsov** [comment]
> Would you happen to have a test project ?

I'll try to prepare it soon.

**Comment by brandelune** [comment]
It looks like it works. Thank you!

I'd suggest:

`Go to **Segment** with Automatic Translation`

_singular_ because you only go to one segment at a time

and

remove `Segment` from the sub-items because it's redundant with the main menu.

What do you think ?

**Comment by brandelune** [comment]
It builds locally, but when I run ./gradlew check I get a failure, like the CI check above:

```
> Task :compileTestJava FAILED
/Users/suzume/Documents/Repositories/omegat.brandelune/test/src/org/omegat/core/TestCore.java:318: error: <anonymous org.omegat.core.TestCore$4> is not abstract and does not override abstract method prevXEnforcedEntry() in IEditor
        TestCoreInitializer.initEditor(new IEditor() {
                                                     ^
1 error

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 20s
```

**Comment by miurahr** [comment]
> @miurahr do you have any idea ?

It is because @kosivantsov changes an API (IEditor class) but did not update a test code accordingly.

**Comment by miurahr** [comment]
```
Index: test/src/org/omegat/core/TestCore.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/src/org/omegat/core/TestCore.java b/test/src/org/omegat/core/TestCore.java
--- a/test/src/org/omegat/core/TestCore.java	(revision 5c55bac19545a9b658712d2d4b76ccf11e456a7e)
+++ b/test/src/org/omegat/core/TestCore.java	(revision cedbcdc6d63adfb254e09380a9444cf6c636c6f3)
@@ -393,6 +393,22 @@
             public void prevEntry() {
             }
 
+            @Override
+            public void nextXAutoEntry() {
+            }
+
+            @Override
+            public void prevXAutoEntry() {
+            }
+
+            @Override
+            public void nextXEnforcedEntry() {
+            }
+
+            @Override
+            public void prevXEnforcedEntry() {
+            }
+
             @Override
             public void nextUntranslatedEntry() {
             }

```

**Review by miurahr — CHANGES_REQUESTED** [review]
Please update test code

**Comment by miurahr** [comment]
Please rebased ONLY topic commits on CURRENT master head. please don't mix another commits.

**Comment by miurahr** [comment]
@kosivantsov  please fix SpotBugs error. SpotBugs is static souce code analyzer and its error means your program code has a bug.

You can run SpotBugs with `./gradlew spotbugsMain` 
then you will find a message like 

```
Execution failed for task ':spotbugsMain'.
> A failure occurred while executing com.github.spotbugs.snom.internal.SpotBugsRunnerForWorker$SpotBugsExecutor
   > 1 SpotBugs violations were found. See the report at: file:///home/kos/omegat/build/reports/spotbugs/main.html
```

Then read report on web browser.

File: src/org/omegat/gui/editor/EditorController.java [hunk 1]
```diff
@@ -1456,6 +1456,40 @@ public void nextTranslatedEntry() {
         nextTranslatedEntry(true);
     }
 
+    private void linkedEntry(boolean forward, String linked) {
+        iterateToEntry(forward, ste -> {
+                TMXEntry info = Core.getProject().getTranslationInfo(ste);
+                if (String.valueOf(info.linked) == linked) {
```
Comment (miurahr) [inline src/org/omegat/gui/editor/EditorController.java:1462]: What are you doing here?

```
                if (String.valueOf(info.linked) == linked) {
```

File: src/org/omegat/gui/editor/IEditor.java [hunk 2]
```diff
@@ -158,6 +158,34 @@ public static CaretPosition startOfEntry() {
      */
     void prevEntry();
 
+    /**
+     * Move to next x-auto translated entry
```
Comment (miurahr) [inline src/org/omegat/gui/editor/IEditor.java:162]: need period for javadoc.

File: src/org/omegat/gui/editor/IEditor.java [hunk 3]
```diff
@@ -158,6 +158,34 @@ public static CaretPosition startOfEntry() {
...
+     *
+     * Must be called only from UI thread.
+     */
+    void nextXAutoEntry();
+
+    /**
+     * Move to previous x-auto translated entry
```
Comment (miurahr) [inline src/org/omegat/gui/editor/IEditor.java:169]: '.'

File: src/org/omegat/gui/editor/IEditor.java [hunk 4]
```diff
@@ -158,6 +158,34 @@ public static CaretPosition startOfEntry() {
...
+     *
+     * Must be called only from UI thread.
+     */
+    void prevXAutoEntry();
+
+    /**
+     * Move to next x-enforced translated entry
```
Comment (miurahr) [inline src/org/omegat/gui/editor/IEditor.java:176]: '.'

File: src/org/omegat/gui/editor/IEditor.java [hunk 5]
```diff
@@ -158,6 +158,34 @@ public static CaretPosition startOfEntry() {
...
+     *
+     * Must be called only from UI thread.
+     */
+    void nextXEnforcedEntry();
+
+    /**
+     * Move to previous x-enforced translated entry
```
Comment (miurahr) [inline src/org/omegat/gui/editor/IEditor.java:183]: '.'

**Review by miurahr — CHANGES_REQUESTED** [review]

File: src/org/omegat/gui/editor/EditorController.java (hunk 1 above)
Comment (miurahr) [reply src/org/omegat/gui/editor/EditorController.java:1462]: Where is a test code of this? Please put unit test for the additions.
I cannot understand why it works correctly.

Comment (kosivantsov) [reply src/org/omegat/gui/editor/EditorController.java:1462]: I have very little idea about tests, but I'll try to look into it. I don't have time till the end of the week, but I'll try to resolve it asap.

Comment (miurahr) [reply src/org/omegat/gui/editor/EditorController.java:1462]: > @kosivantsov please fix SpotBugs error. SpotBugs is static souce code analyzer and its error means your program code has a bug.
> 
> You can run SpotBugs with `./gradlew spotbugsMain` then you will find a message like
> 
> ```
> Execution failed for task ':spotbugsMain'.
> > A failure occurred while executing com.github.spotbugs.snom.internal.SpotBugsRunnerForWorker$SpotBugsExecutor
>    > 1 SpotBugs violations were found. See the report at: file:///home/kos/omegat/build/reports/spotbugs/main.html
> ```
> 
> Then read report on web browser.

FYI. A common idiom for conditionals of null-able string in OmegaT is `if (!StringUtil.isEmpty(a) && a.equals(b)) {...}`. 

I don't know why you need to use a special equal operator "==" , and it becomes a fatal error on SpotBugs static code analysis and quality assurance tool.

**Comment by brandelune** [comment]
@kosivantsov is there anything you need help with here? I'll gladly take a look if it's within my abilities 😄

**Comment by msoutopico** [comment]
> It looks like it works. Thank you!
> 
> I'd suggest:
> 
> `Go to **Segment** with Automatic Translation`
> 
> _singular_ because you only go to one segment at a time
> 
> and
> 
> remove `Segment` from the sub-items because it's redundant with the main menu.
> 
> What do you think ?

I suppose you mean translated with MT? If that's what you mean, how does OmegaT know that the translation came from MT?

**Comment by brandelune** [comment]
> I suppose you mean translated with MT? If that's what you mean, how does OmegaT know that the translation came from MT?

No. It's segments that were automatically inserted from tm/auto/ or tm/enforce/.

**Comment by msoutopico** [comment]
> > I suppose you mean translated with MT? If that's what you mean, how does OmegaT know that the translation came from MT?
> 
> No. It's segments that were automatically inserted from tm/auto/ or tm/enforce/.

Ah right, you're talking about the wording (singular), sorry I misread. Yes, singular for "segment" makes sense, you go to them one by one. 

However, does your suggestion apply to _any_ auto-populated translation coming from either `/tm/auto` and `/tm/enforce` (without distinction)? We need the two separate, but we could have a third one that combines those two.

Not sure what Kos called the two new functions, but my suggestion would be: 
* Go to next auto-populated segment
* Go to next auto-populated locked segment

Or (a bit longer):
* Go to next segment auto-populated with automatic match
* Go to next segment auto-populated with enforced match

although I think "automatic translation" can be misleading...

**Comment by brandelune** [comment]
> although I think "automatic translation" can be misleading...

That's my thought, too. "auto-populated" sounds more accurate.

Regarding splitting the function based on the origin, that would probably be a nice to have item, if @kosivantsov finds that trivial to implement. But that can also be something we add after in a follow-up release.

**Comment by msoutopico** [comment]
> > although I think "automatic translation" can be misleading...
> 
> That's my thought, too. "auto-populated" sounds more accurate.

Agreed! :)
 
> Regarding splitting the function based on the origin, that would probably be a nice to have item, if @kosivantsov finds that trivial to implement. But that can also be something we add after in a follow-up release.

My understanding is that his patch includes _both_ functions (for x-auto and x-enforce).

**Comment by brandelune** [comment]
> My understanding is that his patch includes _both_ functions (for x-auto and x-enforce).

Ooops. You're right.

Also, it looks like the PR includes things that are no related to the function (text modifications to the manual)...

**Comment by miurahr** [comment]
@kosivantsov please rebase to master, fix the bug, make it pass against all tests, and adjust review comments?
You are recommended also to run `./gradlew spotlessChangedApply` linting. see [Developer manual](https://omegat.readthedocs.io/en/latest/30.CodingStyles/)

I don't merge a change with a bug.

**Comment by brandelune** [comment]
I think @kosivantsov is very busy at the moment. I'll see what I can do tomorrow on that issue. @Kazephil if you have the time to take a look, don't hesitate ;-)

**Comment by Kazephil** [comment]
> @Kazephil if you have the time to take a look, don't hesitate ;-)

I'll take a look and see if I can do anything about it.

**Comment by kosivantsov** [comment]
What if we close this PR, and I'll open a new clean one with all the things above addressed?

**Comment by kosivantsov** [comment]
Made a PR #370 from a clean branch with the requested changes.

**Comment by kosivantsov** [comment]
> I think @kosivantsov is very busy at the moment. I'll see what I can do tomorrow on that issue. @Kazephil if you have the time to take a look, don't hesitate ;-)

I gave it another try: #370

**Comment by kosivantsov** [comment]
> @kosivantsov please rebase to master, fix the bug, make it pass against all tests, and adjust review comments? You are recommended also to run `./gradlew spotlessChangedApply` linting. see [Developer manual](https://omegat.readthedocs.io/en/latest/30.CodingStyles/)
> 
> I don't merge a change with a bug.

I really appreciate your comments and guidance, Hiroshi. Now I've run a few checks that you've suggested here and in the linked doc, all of them passed locally.

**Pull request closed without merging**